### PR TITLE
feat: smarter workspace config

### PR DESCRIPTION
**Pull request by jquense**
ensure's that `composite` is set and adds the option to generate build tsconfigs

### CODE REVIEW

File: packages/ts-doctor/README.md [hunk 1]
```diff
@@ -1,3 +1,63 @@
...
+to watch and build each package in order to have up to date type information everywhere.
+In addition editor tooling should show source files not generated
+`.d.ts` files, for "go to definition" and intellisense. Project references allow
+for this, but come at the cost of a lot of arcane setup. Luckily `ts-doctor` can automate
+the vast majority of it!
+
+There are two prequisites for running the command.
+
+1. Your workspaces need to be defined defined. Follow the instructions for your tool of choice, we'll use yarn workspaces in the example.
+2. Each package in your monorepo that should have it's own `tsconfig.json`. This is how the command knows which
```
Comment (itajaja) [inline packages/ts-doctor/README.md]:
```suggestion
2. Each package in your monorepo that should have its own `tsconfig.json`. This is how the command knows which
```

File: packages/ts-doctor/workspaces.js [hunk 2]
```diff
@@ -17,17 +18,56 @@ exports.builder = _ =>
   _.option('cwd', {
     type: 'path',
     describe: 'The current working directory',
-  });
-
-async function getTsConfig(dir) {
+  })
+    .option('with-build-configs', {
```
Comment (itajaja) [inline packages/ts-doctor/workspaces.js]: should be opt-out?

File: packages/ts-doctor/workspaces.js [hunk 3]
```diff
@@ -17,17 +18,56 @@ exports.builder = _ =>
...
+      type: 'boolean',
+      describe:
+        'Creates a tsconfig.build.json file in each package configured for buildind .d.ts files per package. ' +
+        'Useful for Babel based workflows, where `tsc` is only used to build type defs, not compile source. ' +
+        'This is unfortunately necessary because of how compiler flags interact badly with composite projects, making ' +
+        'it impossible to build _just_ type definitions from the root.',
+    })
+    .option('with-sources-metadata', {
```
Comment (itajaja) [inline packages/ts-doctor/workspaces.js]: same here?

**Comment by itajaja** [comment]
q: why isn't ts-doctor part of `cli`?

File: packages/ts-doctor/workspaces.js (hunk 3 above)
Comment (jquense) [reply packages/ts-doctor/workspaces.js]: this is fairly optional i think, not a lot of stuff uses right now.

File: packages/ts-doctor/workspaces.js [hunk 4]
```diff
@@ -68,8 +107,41 @@ function buildWorkspaceSources(baseDir, tsPackages) {
 
   return sources;
 }
+async function resolveTsConfig(dir, configFile = 'tsconfig.json') {
```
Comment (taion) [inline packages/ts-doctor/workspaces.js]:
```suggestion

async function resolveTsConfig(dir, configFile = 'tsconfig.json') {
```

File: packages/ts-doctor/workspaces.js [hunk 5]
```diff
@@ -17,17 +18,56 @@ exports.builder = _ =>
...
+    });
+
+async function getTsConfig(dir, configFile = 'tsconfig.json') {
+  const file = `${dir}/${configFile}`;
   let tsConfigStr;
   try {
-    tsConfigStr = await fs.readFile(`${dir}/tsconfig.json`, 'utf-8');
+    tsConfigStr = await fs.readFile(file, 'utf-8');
   } catch {
+    console.log(file, 'HI');
```
Comment (taion) [inline packages/ts-doctor/workspaces.js]: ?

File: packages/ts-doctor/workspaces.js [hunk 6]
```diff
@@ -68,8 +107,41 @@ function buildWorkspaceSources(baseDir, tsPackages) {
...
+    dir,
+  ).options;
+
+  return { tsConfig, compilerOptions };
+}
+
+async function writeBuildConfig(dir) {
+  const configFile = 'tsconfig.build.json';
+
+  const config = await resolveTsConfig(dir, configFile);
```
Comment (taion) [inline packages/ts-doctor/workspaces.js]: i think you just want `getTsConfig` here – you don't use the calculated `compilerOptions`

File: packages/ts-doctor/workspaces.js (hunk 5 above)
Comment (jquense) [inline packages/ts-doctor/workspaces.js]:
```suggestion
```

File: packages/ts-doctor/workspaces.js [hunk 7]
```diff
@@ -17,17 +18,56 @@ exports.builder = _ =>
...
   } catch {
+    console.log(file, 'HI');
     return null;
   }
 
-  return commentJson.parse(tsConfigStr);
+  try {
+    return commentJson.parse(tsConfigStr);
+  } catch (err) {
+    console.log(tsConfigStr);
```
Comment (jquense) [inline packages/ts-doctor/workspaces.js]:
```suggestion
```

**Pull request merged**